### PR TITLE
agent,plugin: Refer to memory quantities in bytes, not memory slots

### LINKED ISSUE #590: Cleanup: Scheduler should refer to memory in bytes, not memory slots
## Problem description / Motivation

Currently the scheduler plugin measures reserved memory with "memory slots", with the size of a memory slot configured globally in its config.

This is a problem for a few reasons:

1. It means all VMs must have the same memory slot size
2. It limits our flexibility for implementing future features, like #517
3. It causes small imprecisions in our scheduling, because allocatable memory is rounded down to the nearest memory slot size and usage from non-VM pods is rounded up.
4. It blocks other simplifications we can make, like unifying our accounting for VM pods and non-VM pods (which would have avoided the need for #585)

## DoD

MVP: Scheduler plugin no longer internally uses memory slots to represent memory quantities

Follow-up changes:

- unified accounting in scheduler plugin of resource usage from VM pods and non-VM pods
- compute unit is stored in autoscaler-agent's config
- communication between autoscaler-agent and scheduler plugin uses bytes, not memory slots (not sure whether this would be beneficial, tbh)

### PR DESCRIPTION
**Pull request by sharnoff**
Ended up bundling a lot of partially related changes into this one commit. Going to split it into multiple logically separate PRs to make things more manageable.

There's definitely some remaining bugs from places I forgot to change, partly because of how wide-reaching this PR is. In some ways, I think I'll need to do a single pass over all of `pkg/{agent,plugin}` to make sure I didn't miss anything.

Things to split out:

- [x] https://github.com/neondatabase/autoscaling/pull/654
- [x] https://github.com/neondatabase/autoscaling/pull/655
- [x] https://github.com/neondatabase/autoscaling/pull/656
- [x] https://github.com/neondatabase/autoscaling/pull/657
- [x] https://github.com/neondatabase/autoscaling/pull/658
- [ ] ~~plugin: Unify logic from `handlePodStarted` and `Reserve` with `reserveResources`~~
    - Split out into https://github.com/neondatabase/autoscaling/pull/666 
- [ ] ~~plugin: Unify logic from `handlePodDeletion`, `handleVMDeletion`, and `Unreserve` with `unreserveResources`~~
    - Split out into https://github.com/neondatabase/autoscaling/pull/666

Planned final set of changes in this PR:

- Switch `api.Resources.Mem` to refer to bytes, not memory slots
    - This _also_ implies updating the agent<->plugin protocol and all the places in the autoscaler-agent and scheduler plugin that use it
- plugin: Represent memory quantities as bytes, not memory slots
- plugin: Unify `otherPods` and `podsMap` into single set of pods for VMs and non-VM pods (new name `pods`)
    - the original reason to keep them separate was to allow better marginal accounting to deal with rounding down from memory slots. Now that we're referring to it in bytes, there's no longer a need for that
- plugin: Rename `nodeMap` to `nodes`

Resolves #590.

### CODE REVIEW

**Comment by sharnoff** [comment]
Ok, made some updates to this PR — it's now working, and I've checked locally that it's able to work across rollout (i.e. "new scheduler" + "old agents" passes the scaling test).

For reference, I used:

```sh
git checkout main
# pre-build so cluster setup is easier
make docker-build && make docker-build-examples
make kind-setup && make deploy && make example-vms
# upgrade only the scheduler:
git checkout memory-quantities-as-bytes
make docker-build-scheduler && \
    make render-manifests && \
    make kind-load && \
    kubectl apply -f rendered_manifests/autoscale-scheduler.yaml
# actually test it
make e2e
```

**Review by shayanh — COMMENTED** [review]
I left a few comments and also there is conflict with main. Let's resolve these and then I'll go over the changes again.

File: pkg/api/vminfo.go [hunk 1]
```diff
@@ -59,40 +59,41 @@ type VmCpuInfo struct {
 }
 
 type VmMemInfo struct {
+	// Min is the minimum number of memory slots available
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/api/vminfo.go:62]: aad951c948c925d6b18a72638425a7f3a9fdceeb

File: pkg/api/types.go [hunk 2]
```diff
@@ -144,6 +165,57 @@ func (r AgentRequest) ProtocolRange() VersionRange[PluginProtoVersion] {
...
+func (b Bytes) Format(state fmt.State, verb rune) {
+	switch {
+	case verb == 'v' && state.Flag('#'):
+		state.Write([]byte(fmt.Sprintf("%v", uint64(b))))
+	default:
+		state.Write([]byte(b.ToResourceQuantity().String()))
+	}
+}
+
 // Resources represents an amount of CPU and memory (via memory slots)
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/api/types.go]: 0ecbb297580f03e0afc09082a7a78c0ef5d4186b

File: pkg/api/types.go [hunk 3]
```diff
@@ -58,9 +60,18 @@ const (
 	//
 	// * added AgentRequest.LastPermit
 	//
-	// Currently the latest version.
+	// Last used in release version <TODO>.
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/api/types.go:63]: Yep, good catch. Will leave this unresolved until it's closer to merging.

**Comment by sharnoff** [comment]
update: It's probably easier to merge and deploy #707 first.

**Review by shayanh — APPROVED** [review]
I tried to review this PR again more thoroughly this time and looks good.

**Comment by sharnoff** [comment]
Status: #707 has been merged but not yet deployed - waiting on that. Also need to resolve merge conflicts here.

**Comment by shayanh** [comment]
@sharnoff looks good. let's go ahead and merge this.

**Pull request merged**